hack/generator/pkg/codegen/crossplane_pipeline_add_embedded_resource_types.go

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+)
+
+var CrossplaneRuntimeV1Alpha1Package = astmodel.MakeExternalPackageReference("github.com/crossplane/crossplane-runtime/apis/core/v1alpha1")
+
+// addCrossplaneEmbeddedResourceSpec puts an embedded runtimev1alpha1.ResourceSpec on every spec type
+func addCrossplaneEmbeddedResourceSpec(idFactory astmodel.IdentifierFactory) PipelineStage {
+
+	return MakePipelineStage(
+		"addCrossplaneEmbeddedResourceSpec",
+		"Puts an embedded runtimev1alpha1.ResourceSpec on every spec type",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			specTypeName := astmodel.MakeTypeName(
+				CrossplaneRuntimeV1Alpha1Package,
+				idFactory.CreateIdentifier("ResourceSpec", astmodel.Exported))
+			embeddedSpec := astmodel.NewPropertyDefinition("", ",inline", specTypeName)
+
+			result := make(astmodel.Types)
+			for _, typeDef := range types {
+
+				if resource, ok := typeDef.Type().(*astmodel.ResourceType); ok {
+
+					specDef, err := types.ResolveResourceSpecDefinition(resource)
+					if err != nil {
+						return nil, errors.Wrapf(err, "getting resource spec definition")
+					}
+
+					// The assumption here is that specs are all Objects
+					updatedDef, err := specDef.ApplyObjectTransformation(func(o *astmodel.ObjectType) (astmodel.Type, error) {
+						return o.WithEmbeddedProperty(embeddedSpec)
+					})
+					if err != nil {
+						return nil, errors.Wrapf(err, "adding embedded crossplane spec")
+					}
+
+					result.Add(typeDef)
+					result.Add(updatedDef)
+				}
+			}
+
+			for _, typeDef := range types {
+				if !result.Contains(typeDef.Name()) {
+					result.Add(typeDef)
+				}
+			}
+
+			return result, nil
+		})
+}
+
+// addCrossplaneEmbeddedResourceStatus puts an embedded runtimev1alpha1.ResourceStatus on every spec type
+func addCrossplaneEmbeddedResourceStatus(idFactory astmodel.IdentifierFactory) PipelineStage {
+
+	return MakePipelineStage(
+		"addCrossplaneEmbeddedResourceStatus",
+		"Puts an embedded runtimev1alpha1.ResourceStatus on every status type",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			statusTypeName := astmodel.MakeTypeName(
+				CrossplaneRuntimeV1Alpha1Package,
+				idFactory.CreateIdentifier("ResourceStatus", astmodel.Exported))
+			embeddedStatus := astmodel.NewPropertyDefinition("", ",inline", statusTypeName)
+
+			result := make(astmodel.Types)
+			for _, typeDef := range types {
+				if resource, ok := typeDef.Type().(*astmodel.ResourceType); ok {
```

Comment:
I've done this as it's the correct (ish) thing to do now, although there are issues in some instances of this PR where this was suggested because we're modifying a type and throwing away what wrapped it.

I'm doing some other work (that involves flagging all of the status types) and doing that basically breaks _all_ of our pipelines for one reason or another because of the nested types problem. I really think we need to consider alternatives to it - I may schedule a small whiteboard meeting to discuss them at some point after this PR gets merged.